File: backend/src/main/java/org/luxons/sevenwonders/lobby/Player.java
package org.luxons.sevenwonders.lobby;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.luxons.sevenwonders.game.Game;

public class Player {

    private final String username;

    private String displayName;

    private int index;

    private boolean ready;

    private transient Lobby lobby;

    private transient Game game;

    public Player(String username, String displayName) {
        this.username = username;
        this.displayName = displayName;
    }

    public String getUsername() {
        return username;
    }

    public String getDisplayName() {
        return displayName;
    }

    public void setDisplayName(String displayName) {
        this.displayName = displayName;
    }

    public int getIndex() {
        return index;
    }

    public void setIndex(int index) {
        this.index = index;
    }

    public boolean isReady() {
        return ready;
    }

    public void setReady(boolean ready) {
        this.ready = ready;
    }

    @JsonIgnore
    public Lobby getLobby() {
        return lobby;
    }

    public void setLobby(Lobby lobby) {
        this.lobby = lobby;
    }

    @JsonIgnore
    public Game getGame() {
        return game;
    }

    public void setGame(Game game) {
        this.game = game;
    }

    @Override
    public String toString() {
        return "'" + displayName + "' (" + username + ")";
    }
}

File: backend/src/main/kotlin/org/luxons/sevenwonders/lobby/Player.kt
package org.luxons.sevenwonders.lobby

import com.fasterxml.jackson.annotation.JsonIgnore
import org.luxons.sevenwonders.errors.ApiMisuseException
import org.luxons.sevenwonders.game.Game

class Player(
    val username: String,
    var displayName: String
) {
    var index: Int = -1

    var isReady: Boolean = false

    val isGameOwner: Boolean
        get() = _lobby?.isOwner(username) ?: false

    val isInLobby: Boolean
        get() = _lobby != null

    val isInGame: Boolean
        get() = _game != null

    @Transient
    private var _lobby: Lobby? = null

    @get:JsonIgnore
    val lobby: Lobby
        get() = _lobby ?: throw PlayerNotInLobbyException(username)

    @get:JsonIgnore
    val ownedLobby: Lobby
        get() = if (isGameOwner) lobby else throw PlayerIsNotOwnerException(username)

    @Transient
    private var _game: Game? = null

    @get:JsonIgnore
    val game: Game
        get() = _game ?: throw PlayerNotInGameException(username)

    fun join(lobby: Lobby) {
        _lobby = lobby
    }

    fun join(game: Game, index: Int) {
        _game = game
        this.index = index
    }

    fun leave() {
        _lobby = null
        _game = null
        index = -1
    }

    override fun toString(): String {
        return "'$displayName' ($username)"
    }
}

internal class PlayerNotInLobbyException(username: String) :
    ApiMisuseException("User $username is not in a lobby, create or join a game first")

internal class PlayerIsNotOwnerException(username: String) :
    ApiMisuseException("User $username does not own the lobby he's in")

internal class PlayerNotInGameException(username: String) :
    ApiMisuseException("User $username is not in a game, start a game first")
